core/crypto/secp256k1/impl/secp256k1_provider_impl.cpp

```diff
@@ -0,0 +1,146 @@
+    if (BN_bn2binpad(privateNum, private_key.data(), private_key.size()) < 0) {
+      return Secp256k1Error::KEY_GENERATION_FAILED;
+    }
+    EC_KEY_set_conv_form(key.get(), POINT_CONVERSION_UNCOMPRESSED);
+    auto public_key_length = i2o_ECPublicKey(key.get(), nullptr);
+    if (public_key_length != public_key.size()) {
+      return Secp256k1Error::KEY_GENERATION_FAILED;
+    }
+    uint8_t *public_key_ptr = public_key.data();
+    if (i2o_ECPublicKey(key.get(), &public_key_ptr) != public_key_length) {
+      return Secp256k1Error::KEY_GENERATION_FAILED;
+    }
+    return KeyPair{private_key, public_key};
+  }
+
+  outcome::result<PublicKeyUncompressed> Secp256k1ProviderImpl::derive(
+      const PrivateKey &key) const {
+    secp256k1_pubkey pubkey;
+
+    if (!secp256k1_ec_pubkey_create(context_.get(), &pubkey, key.data())) {
+      return Secp256k1Error::KEY_GENERATION_FAILED;
+    }
+
+    PublicKeyUncompressed public_key{};
+    size_t outputlen = kPublicKeyUncompressedLength;
+    if (!secp256k1_ec_pubkey_serialize(context_.get(),
+                                       public_key.data(),
+                                       &outputlen,
+                                       &pubkey,
+                                       SECP256K1_EC_UNCOMPRESSED)) {
+      return Secp256k1Error::PUBKEY_SERIALIZATION_ERROR;
+    }
+
+    return public_key;
+  }
+
+  outcome::result<PublicKeyUncompressed> Secp256k1ProviderImpl::sign(
+      gsl::span<const uint8_t> message, const PrivateKey &key) const {
+    secp256k1_ecdsa_recoverable_signature sig_struct;
+    if (!secp256k1_ecdsa_sign_recoverable(context_.get(),
+                                          &sig_struct,
+                                          message.data(),
+                                          key.cbegin(),
+                                          secp256k1_nonce_function_rfc6979,
+                                          nullptr)) {
+      return Secp256k1Error::CANNOT_SIGN_ERROR;
+    }
+    SignatureCompact signature{};
+    int recid = 0;
+    if (!secp256k1_ecdsa_recoverable_signature_serialize_compact(
+            context_.get(), signature.data(), &recid, &sig_struct)) {
+      return Secp256k1Error::SIGNATURE_SERIALIZATION_ERROR;
+    }
+    signature[64] = (uint8_t)recid;
+    return signature;
+  }
+
+  outcome::result<bool> Secp256k1ProviderImpl::verify(
+      gsl::span<const uint8_t> message,
+      const SignatureCompact &signature,
+      const PublicKeyUncompressed &key) const {
+    OUTCOME_TRY(checkSignature(signature));
+
+    secp256k1_ecdsa_signature sig;
+    secp256k1_pubkey pubkey;
+
+    if (!secp256k1_ecdsa_signature_parse_compact(
+            context_.get(), &sig, signature.data())) {
+      return Secp256k1Error::SIGNATURE_PARSE_ERROR;
+    }
+    if (!secp256k1_ec_pubkey_parse(
+            context_.get(), &pubkey, key.data(), key.size())) {
+      return Secp256k1Error::PUBKEY_PARSE_ERROR;
+    }
+    return secp256k1_ecdsa_verify(
```

Comment:
does it return bool? if not, pls static cast it or convert other way

Reply:
It returns 1 in case of success, or 0 otherwise. Added check.